### PR TITLE
❄ chore: Add nix flake

### PR DESCRIPTION
**Pull request by shazow**
[##](url) Description

image

## Testing

`nix develop` to enter the flake environment, confirmed bun/pnpm/rust/etc are present.

## Additional Information

️- [ ] I read the [contributing docs](../docs/contributing.md) (if this is your first contribution)
  **path is borken ⬆**

Your ENS/address: shazow.eth

## Summary by CodeRabbit

- **New Features**
	- Introduced a new development shell configuration for streamlined project setup.
	- Added essential tools including Node.js, pnpm, bun, Foundry, and Rust to the development environment.
	- Enhanced user experience with a customizable shell prompt and environment file sourcing.

- **Refactor**
	- Organized dependency management through structured inputs from various repositories.

### CODE REVIEW

File: flake.lock [hunk 1]
```diff
@@ -0,0 +1,176 @@
+{
+  "nodes": {
```
Comment (roninjin10) [inline flake.lock:2]: Each entry under "nodes" represents a dependency or a part of the project's dependency graph.

File: flake.lock [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+      },
+      "original": {
+        "owner": "nix-systems",
+        "repo": "default",
+        "type": "github"
+      }
+    }
+  },
+  "root": "root",
+  "version": 7
```
Comment (roninjin10) [inline flake.lock:175]: Indicates the version of the lockfile format

File: flake.lock [hunk 3]
```diff
@@ -0,0 +1,176 @@
+{
+  "nodes": {
+    "flake-parts": {
+      "inputs": {
```
Comment (roninjin10) [inline flake.lock:4]: Specifies other nodes this node depends on.

File: flake.lock [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+      },
+      "locked": {
+        "lastModified": 1722555600,
+        "narHash": "sha256-XOQkdLafnb/p9ij77byFQjDf5m5QYl9b2REiVClC+x4=",
+        "owner": "hercules-ci",
+        "repo": "flake-parts",
+        "rev": "8471fe90ad337a8074e957b69ca4d0089218391d",
+        "type": "github"
+      },
+      "original": {
```
Comment (roninjin10) [inline flake.lock:15]: The original reference used to fetch the dependency

File: flake.lock [hunk 5]
```diff
@@ -0,0 +1,176 @@
+{
+  "nodes": {
+    "flake-parts": {
+      "inputs": {
+        "nixpkgs-lib": "nixpkgs-lib"
+      },
+      "locked": {
+        "lastModified": 1722555600,
+        "narHash": "sha256-XOQkdLafnb/p9ij77byFQjDf5m5QYl9b2REiVClC+x4=",
```
Comment (roninjin10) [inline flake.lock:9]: The NAR (Nix Archive) hash of the dependency, ensuring integrity.

File: flake.lock [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        "owner": "hercules-ci",
```
Comment (roninjin10) [inline flake.lock:10]: owner: GitHub repository owner.

File: flake.lock [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        "repo": "flake-parts",
```
Comment (roninjin10) [inline flake.lock:11]: repo: Repository name.

File: flake.lock [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        "rev": "8471fe90ad337a8074e957b69ca4d0089218391d",
```
Comment (roninjin10) [inline flake.lock:12]: rev: The specific git commit hash that is locked.

File: flake.lock [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        "type": "github"
```
Comment (roninjin10) [inline flake.lock:13]: type: Type of source, e.g., "github" or "tarball".

File: flake.nix [hunk 10]
```diff
@@ -0,0 +1,37 @@
+# `nix develop` to enter the devShell
+{
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+    flake-parts.url = "github:hercules-ci/flake-parts";
+    foundry.url = "github:shazow/foundry.nix/monthly";
```
Comment (roninjin10) [inline flake.nix:6]: Have you asked foundry if they wanted to upstream this?

File: flake.nix [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    rust-overlay.url = "github:oxalica/rust-overlay";
+  };
+
+  outputs = { nixpkgs, flake-utils, foundry, rust-overlay, ... }:
```
Comment (roninjin10) [inline flake.nix:10]: The outputs section defines the configurations for the development environment:

File: flake.nix [hunk 12]
```diff
@@ -0,0 +1,37 @@
+# `nix develop` to enter the devShell
+{
+  inputs = {
```
Comment (roninjin10) [inline flake.nix:3]: The inputs section defines the sources for the dependencies:

File: flake.nix [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    flake-utils.lib.eachDefaultSystem (system:
+      let pkgs = import nixpkgs {
+        inherit system;
+        overlays = [
+          foundry.overlay
+          rust-overlay.overlays.default
+        ];
+      };
+    in {
+      devShells.default = pkgs.mkShell {
```
Comment (roninjin10) [inline flake.nix:20]: Defines the default development shell environment for the project.

File: flake.nix [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+        buildInputs = [
```
Comment (roninjin10) [inline flake.nix:21]: Specifies the packages to be included in the development shell:

File: flake.nix [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+          # "if you add npm 22 bun pnpm foundry and rust nix configs to Tevm repo I’ll start using and shilling nix" ~ fucory.eth
+          pkgs.nodejs_22
+          pkgs.nodePackages.pnpm
+          pkgs.bun
+          pkgs.foundry-bin
+          pkgs.rust-bin.stable.latest.default # or pkgs.rust-bin.beta.latest.default
+        ];
+
+        shellHook = ''
```
Comment (roninjin10) [inline flake.nix:30]: A script executed each time the development shell is started

File: flake.nix [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+          export PS1="[dev] $PS1"
+
+          [[ -f .env ]] && source .env
```
Comment (roninjin10) [inline flake.nix:33]: oh this is nice before for a small handful of tests there was no good way for outside contributors to set the env variables yet so this fixes a long standing TODO

File: flake.nix [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+          # "if you add npm 22 bun pnpm foundry and rust nix configs to Tevm repo I’ll start using and shilling nix" ~ fucory.eth
+          pkgs.nodejs_22
+          pkgs.nodePackages.pnpm
+          pkgs.bun
+          pkgs.foundry-bin
+          pkgs.rust-bin.stable.latest.default # or pkgs.rust-bin.beta.latest.default
+        ];
+
+        shellHook = ''
+          export PS1="[dev] $PS1"
```
Comment (roninjin10) [inline flake.nix:31]: export PS1="[dev] $PS1": Modifies the shell prompt to indicate the dev shell.

**Review by roninjin10 — APPROVED** [review]
Very nice!  Couple minor questions but mostly left a bunch of comments for my own reference

**Comment by roninjin10** [comment]
follow up issue https://github.com/evmts/tevm-monorepo/issues/1366

File: flake.lock (hunk 1 above)
Comment (shazow) [reply flake.lock:2]: FWIW the `flake.lock` file is generated by running `nix flake update` to update to latest (or `nix flake lock` to re-pin current versions). I generally don't audit it independently unless I'm trying to debug something version-related.

File: flake.nix (hunk 16 above)
Comment (shazow) [reply flake.nix:33]: Yea feel free to add whatever in here.

Sometimes I'll have things like:

```bash
          export PATH=$PWD/node_modules/.bin:$PATH

          [[ ! -d node_modules ]] && pnpm install
```

**Pull request merged**